align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+            # If there is only one, make it a list for the following loop
+            if not isinstance(dset_files, list):
+                dset_files = [dset_files]
+            
+            # For each dataset file
+            for dset_f in dset_files:
+                with open(dset_f) as f:
+                    dset = json.load(f)
+                incontext_data[sys_kdma_name] = []
+                # Load each example in the dataset file
+                for icl_sample in dset:
+                    # Get state and actions
+                    state, actions = hydrate_scenario_state(icl_sample["input"])
+                    # Get choices
+                    choices = adm_utils.format_choices(
+                        [a.unstructured for a in actions],
+                        actions,
+                        state,
+                        self.log
+                    )
+                    # Get KDMA_values
+                    kdma_values = []
+                    for label in icl_sample["label"]:
+                        if sys_kdma_name not in label:
+                            kdma_values.append(None)
+                        else:
+                            kdma_values.append(label[sys_kdma_name])
+                    example = {'state':state, 'actions': actions, 'choices':choices, 'kdma_values':kdma_values}
+                    incontext_data[sys_kdma_name].append(example)
+            
+            # TODO - add KDMA normalization option
+            
+            return incontext_data
+
+    def select_icl_examples(self, sys_kdma_name, prompt_to_match):
+        '''
+        Selects a list of relevant ICL examples
+        Input:
+            sys_kdma_name - key of the target kdma in self.icl_datasets
+            prompt_to_match - the prompt we are selecting ICL examples for
+        Output:
+            selected_icl_examples - relevant subset of self.icl_datasets
+        '''
+        # Check that we have incontext examples for the target kdma
+        if sys_kdma_name not in self.icl_datasets:
+            raise RuntimeError(f"No incontext samples for targeted kdma: {sys_kdma_name}")
+        n_icl_examples = self.incontext_settings["number"]
+        possible_icl_examples = self.icl_datasets[sys_kdma_name]
+        # Check that we have enough incontext examples for the target kdma
+        if len(possible_icl_examples) < n_icl_examples:
+            raise RuntimeError(f"Not enough possible incontext samples to learn from. Only "
+                            f"{len(possible_icl_examples)} samples available while asking for "
+                            f"{n_icl_examples} incontext samples.")
+        # If using LOO, don't include example ICL with exact same prompt
+        if self.incontext_settings.get("leave_one_out", False):
+            possible_icl_examples = [
+                icl_ex for icl_ex in possible_icl_examples
+                if icl_ex["prompt"] != prompt_to_match
```

Comment:
There is a use case where the prompt we check for LOO might be different than the prompt for similarity, e.g we may want to leave out the actions for LOO so that the ordering or a pre-filtering doesn't cause the matching to fail. I'm wondering if @dmjoy (or @jadie1!) has ideas for better matching strategies since this is pretty brittle anyway (yes, I know I originally wrote this :smile: )

Reply:
That is a good point that the choice order or choice variations could cause this to fail. I guess we could just use the scenario description to check? That would require saving the scenario descriptions in addition to the prompts and responses in `self.icl_datasets` and then also passing the "scenario_description_to_match" in `select_icl_examples()`

Reply:
I switched it to just use the scenario description here https://github.com/ITM-Kitware/align-system/commit/f4fdf9060b9ef0d5498e5ee691f060bdd0469f56
Let me know if this is better

Reply:
I mentioned this on the call, but wanted to make it clearer: 

I think this is a good change, but it may not be general enough. If we keep strings in the dataset, we might want a "similarity_prompt" and a "loo_prompt" in addition to the final prompt in our dataset, that way it could be more general. In this particular function, I think that would be just renaming variables/keys, but would require slightly different dataset loading/generation

This is where I think @dmjoy's  comment about specifying functions for generating these prompts instead of strings would be really useful, but I can't picture what that would look like

Not something I would hold up this PR for though